apps/api/src/course/course.service.ts

```diff
@@ -93,26 +80,129 @@ export class CourseService {
           message: 'Start time or end time is invalid',
         })
       }
+
       if (start > end) {
         throw new BadRequestException({
           reason: 'INVALID_PERIOD_RANGE',
           message: 'Start time cannot be later than end time',
         })
       }
-      query['sections.classes'] = {
-        $elemMatch: {
-          'period.start': { $gte: start },
-          'period.end': { $lte: end },
-        },
-      }
     }
-
-    const courses = await this.courseModel.find(query).limit(limit).skip(offset).lean()
-    return courses
+    return this.searchService.search<ICourseSearchDocument>({
+      index: this.configService.get<string>('courseIndex'),
+      query: buildCourseQuery({
+        keyword,
+        genEdTypes,
+        dayOfWeeks,
+        periodRange,
+        studyProgram,
+        semester,
+        academicYear,
+      }),
+      from: offset,
+      size: limit,
+    })
   }
 }
 
-function isTime(timeString: string): boolean {
-  const timeRegex = /^\d{2}:\d{2}$/
-  return timeRegex.test(timeString)
+// build the query
+function buildCourseQuery(filter: ICourseSearchFilter): QueryDslQueryContainer {
```

Comment:
Can you explain what this entire query mean?

Reply:
ยากละ 🥲

Reply:
หลักๆก็คือ function นี้มีไว้เพื่อสร้าง query สำหรับ search course โดยจะ base on [Boolean query](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-bool-query.html)

ส่วนนี้คือ query ที่การันตีว่ามี value จาก input มาแน่นอน
- semester, studyProgram, academicYear ให้เช็คแบบ [Term query](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-term-query.html) (Exact match)
- keyword ให้เช็คแบบ [Multimatch query](https://www.elastic.co/guide/en/elasticsearch/reference/8.6/query-dsl-multi-match-query.html#query-dsl-multi-match-query) เพราะมันอาจจะ hit ได้จากหลายๆ fields โดยที่จะให้ค่าความสำคัญในการ ranking ตาม `^` ที่ใส่อยู่หลังชื่อ field
```ts
  const boolMust: QueryDslQueryContainer[] = [
    {
      multi_match: {
        query: filter.keyword,
        fields: [
          'abbrName^5',
          'courseNo^5',
          'courseNameEn^3',
          'courseDescEn',
          'courseNameTh^3',
          'courseDescTh',
        ],
      },
    },
    {
      term: {
        semester: {
          value: filter.semester,
        },
      },
    },
    {
      term: {
        studyProgram: {
          value: filter.studyProgram,
        },
      },
    },
    {
      term: {
        academicYear: {
          value: filter.academicYear,
        },
      },
    },
  ]
```

ส่วนนี้คือ filter อาจจะเป็น undefined ได้ เลยต้องมาเช็คก่อน push value เข้า query อันที่แล้ว
- dayOfWeek มันเป็น nested data (อยู่ใน rawData) จาก schema `ICourseSearchDocument` เลยต้องใช้ [Nested query](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-nested-query.html) ในการ search คนหา field + ใช้ [Terms query](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-terms-query.html) ในการหา value ที่ match กับของใน array
- periodRange ก็ case เดียวกันแต่ใช้ [Query string query](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-query-string-query.html) เพราะ [Range search](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-range-query.html) ใน Elasticsearch ไม่รองรับ format เวลา
- genEdType case เดียวกับ dayOfWeek
- if อันสุดท้ายคือถ้ามันมีเงื่อนไขที่ต้องใช้ nested query ก็ให้เพิ่มไปใน boolean query อันแรกด้วย
```ts
  // push the query for each filter if filter is not undefined
  if (filter.dayOfWeeks.length > 0) {
    nestedQuery.push({
      terms: {
        'rawData.sections.classes.dayOfWeek': filter.dayOfWeeks,
      },
    })
  }

  if (filter.periodRange) {
    nestedQuery.push({
      nested: {
        path: 'rawData.sections.classes.period',
        query: {
          query_string: {
            query: `rawData.sections.classes.period.start:[${filter.periodRange.start} TO ${filter.periodRange.end}] AND rawData.sections.classes.period.end:[* TO ${filter.periodRange.end}]`,
          },
        },
      },
    })
  }

  if (filter.genEdTypes.length > 0) {
    boolMust.push({
      terms: {
        genEdType: filter.genEdTypes,
      },
    })
  }

  if (nestedQuery.length > 0) {
    boolMust.push({
      nested: {
        path: 'rawData',
        query: {
          nested: {
            path: 'rawData.sections',
            query: {
              nested: {
                path: 'rawData.sections.classes',
                query: {
                  bool: {
                    must: nestedQuery,
                  },
                },
              },
            },
          },
        },
      },
    })
  }
```